File: LC0274_H_Index/h_index.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        
        // trivial case
        if (n == 1)
        {
            return citations[0] == 0 ? 0 : 1;
        }
        sort(citations.begin(), citations.end());
        int temp = 0;
        int h_index = 0;
        for (int i = 0; i < n; i++)
        {
            if (citations[i] == 0)
            {
                temp = 0;
            }
            else
            {
                temp = min(citations[i], n - i);
            }
            h_index = max(h_index, temp);
        }

        return h_index;
    }
};

File: LC0274_H_Index/h_index.py
from typing import List

class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        # trivial cases
        if n == 1:
            if citations[0] == 0:
                return 0
            else:
                return 1

        citations = sorted(citations)
        # For a paper with x citations, check if there are at least x papers, each of which has at least x citations.
        h_index = 0
        temp = 0
        for i in range(n):
            if citations[i] == 0:
                temp = 0
            else:
                temp = min(citations[i], n - i)
            h_index = max(h_index, temp)
        return h_index